Introduce LWM2M subsystem for Zephyr

mike-scott opened the pull request:
Hello all,

I'm submitting this series as RFC to start discussion / early review process to get an LWM2M implementation merged into Zephyr mainline.

A bit of history:
After doing quite a bit of research to select a good starting point for a Zephyr-based LWM2M stack, I selected the following source: https://github.com/sics-iot/lwm2m-contiki/tree/lwm2m-standalone-dtls/apps/oma-lwm2m
Created by:  Joakim Eriksson, Niclas Finne and Joel Hoglund as a stand-alone and/or Contiki based solution.   The stack they've put together is very robust and looked to be a good candidate.

Here's a small summary of my changes from the original sources:
- [engine] The engine portion was re-written due to the heavy reliance on ER-CoAP APIs which are not compatible to the Zephyr CoAP APIs and the other items below marked with [engine].
- [engine] All LWM2M/IPSO object data is now abstracted into resource data which stores information like the data type, length, some callbacks.  The engine modifies this data directly (or makes callbacks) instead of all of the logic for this living in each object's code.  (This wasn't scaling well as I was implementing changes).
- [engine] Related to the above change, I also added a generic set of getter/setter functions that user applications can call to change the object data instead of having to add getter/setting methods in each object.
- [engine] The original sources shared the engine's context structure quite extensively causing a problem with portability.  I broke up the context into it's individual parts: LWM2M path data, input data and output data and pass only the needed data into each set of APIs (for instance reader / writers only get path data and the related input or output structure).
- [sample] The sample/net/lwm2m_client uses the net app API which will be utilized more in the subsys source in the future
- DTLS / Bootstrap was not included and will be added in a later patch
- Float32/Float64 input / output handling functions will be implemented in a later patch
- JSON output handling probably needs to be debugged more. (TLV is the primary delivery format)
- More IPSO Alliance objects will be implemented as we get further into review
- The Temperature Sensor will be finished as review progresses.  Currently, w/o a Float32 implementation it will return empty values.
- CoAP URI parser functions needs to be added and instead of the sample using CONFIG_NET_APP_PEER_IPV6_ADDR / CONFIG_NET_APP_PEER_IPV4_ADDR values directly.  It will assign these as a coap:// URI to the security object 0 index.  Once the LWM2M client is started, it will parse the values out of index 0 and make the correct connections (port, etc) -- including possible DNS lookup.
- Block transfer is only handled in select situations.  A future patch to the engine will add a tracking mechanism by which larger amounts of data can be queued up for sending and/or delivered to user applications in piecemeal when receiving.

The sample for LWM2M can be run via QEMU by following the instructions here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/subsystems/networking/qemu_setup.rst

(Quick Instructions)
To start with you will want to use Leshan LWM2M demo server:
[terminal #1]
$ wget https://hudson.eclipse.org/leshan/job/leshan/lastSuccessfulBuild/artifact/leshan-server-demo.jar
$ java -jar ./leshan-server-demo.jar
(You can now open a web browser to http://localhost:8080/ to watch connections)

Next, checkout https://github.com/zephyrproject-rtos/net-tools and follow the build instructions referenced above.  After they are built, start up the slip tools:
[terminal #2]
$ ./loop-socat.sh
[terminal #3]
$ sudo ./loop-slip-tap.sh

Now you can cd to the Zephyr sources and build the sample for QEMU target:
$ make -C samples/net/lwm2m_client qemu

You should see the sample app connect to Leshan and register twice (IPv4 connection and IPv6 connection).  Feel free to navigate around Leshan and play with the QEMU LWM2M client.

I look forward to your comments.

tbursztyka commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2473 @@
+		}
+
+		/* match delim or end of string */
+		if (begin >= 0 && (pathstr[i] == delim ||
```
having a lot of for/if/else imbrication makes things hard to read. Though here it's contiki legacy I guess, could be worth to change it so if (<inverse of above condition>) { continue; }

mike-scott replied:
This is my code and I'll fix for readability.

tbursztyka commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2473 @@
+		/* if exact match on obj_inst path, then read all resources */
+		/* if not then read only matching resources */
+
+		if (match_type > MATCH_NONE && obj_inst->resources &&
```
ouch, full of if/while/else/... imbrication below, hard to read. (legacy thing again)

mike-scott replied:
Same.  This is my code and I'll fix for readability.

tbursztyka commented on subsys/net/lib/lwm2m/lwm2m_rw_json.c:
```diff
@@ -0,0 +1,644 @@
+/*
```
Would there be a way for using part of zephyr's json library?

mike-scott replied:
Yes, I have a TODO marked in the header:
"Research using Zephyr JSON lib for json_next_token()"

tbursztyka commented on subsys/net/lib/lwm2m/Kconfig.ipso:
```diff
@@ -0,0 +1,34 @@
+
+if LWM2M_IPSO_SUPPORT
+
+config IPSO_TEMP_SENSOR
```
LWM2M_ prefix is needed on all sub-config options in this menu.

mike-scott replied:
Done.

tbursztyka commented on include/misc/byteorder.h:
```diff
@@ -21,6 +21,14 @@
 				   (((x) >> 8) & 0xff00) | \
 				   (((x) & 0xff00) << 8) | \
 				   (((x) & 0xff) << 24)))
+#define __bswap_64(x) ((u64_t) ((((x) >> 56) & 0xff) | \
```
ok but why is this patch part of lwm2m patchset? You probably want to send this patch in another PR.

mike-scott replied:
I added it here so testers could pull this as a part of the PR to test LWM2M.  I will definitely submit this separately.

nashif replied:
as long as it is in its commit, I do not think it make sense to create another PR, the commit is needed here, so it should be part of this PR, lets not go into this please, it is enough to have commits separated and have PRs be more inclusive and allow coverage of major features that touch different parts of the system.

tbursztyka reviewed (changes requested):
legacy code is sometime quite hard to read due to numerous if/for/while/else/... imbrication. I don't know if the plan is to backport future patches from legacy code to this port, if not maybe it will be nice to rearrange the code for better readability.

Could be done later though.

dbkinder commented on doc/subsystems/networking/overview.rst:
```diff
@@ -65,6 +65,12 @@ can be disabled if not needed.
   :ref:`coap-server-sample` using DTLS (Datagram Transport Layer Security)
   (RFC 6347) are also implemented.
 
+* **LWM2M** OMA Lightweight Machine-2-Machine Protocol (V1.0 Feb 2017) is
```
Machine-to-Machine

mike-scott replied:
Done.

dbkinder commented on samples/net/lwm2m_client/README.rst:
```diff
@@ -0,0 +1,19 @@
+.. _lwm2m-client-sample:
+
+LWM2M client
+###########
```
Header underline needs to be at least as long as the title (needs one more #)

mike-scott replied:
Done.

mike-scott commented on include/net/zoap.h:
```diff
@@ -226,6 +226,7 @@ struct zoap_reply {
 	zoap_reply_t reply;
 	void *user_data;
 	int age;
+	u16_t id;
```
Done.  Looks a bit odd to split token / tkl though. :/

mike-scott commented on subsys/net/lib/zoap/zoap.c:
```diff
@@ -614,20 +614,24 @@ struct zoap_reply *zoap_response_received(
 	struct zoap_reply *replies, size_t len)
 {
 	struct zoap_reply *r;
+	u16_t id;
```
Done

mike-scott commented on include/net/lwm2m.h:
```diff
@@ -0,0 +1,155 @@
+#define STATE_DOWNLOADED	2
+#define STATE_UPDATING		3
+
+#define	RESULT_DEFAULT		0
```
This is a file I created, and I don't mind fixing spacing / readability issues like this.  Done.

mike-scott commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -0,0 +1,144 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+menuconfig LWM2M_LIB
```
Heh, I used the MQTT_LIB model.  Removing the _LIB throughout.

mike-scott commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -0,0 +1,144 @@
+config SYS_LOG_LWM2M_LEVEL
+	int "LWM2M log level"
+	depends on SYS_LOG
+	default 3
```
Defaulting to 1.  Done.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2473 @@
+
+#define DISCOVER_PREFACE	"</.well-known/core>;ct=40"
+
+struct observe_node {
```
Done.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2473 @@
+	/* make sure this observer doesn't exist already */
+	SYS_SLIST_FOR_EACH_CONTAINER(&engine_observer_list, obs, node) {
+		if (memcmp(&obs->addr, addr, sizeof(addr)) == 0 &&
+			memcmp(&obs->path, path, sizeof(*path)) == 0) {
```
Adjusting so that the memcmp are lined up vertically.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2473 @@
+	zoap_header_set_version(zpkt, 1);
+	zoap_header_set_type(zpkt, type);
+	zoap_header_set_code(zpkt, code);
+	if (mid > 0) {
```
Done.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2473 @@
+		 * TODO: iterate through resources once block transfer
+		 * is handled correctly
+		 */
+		if (pos + len < size) {
```
Done.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2473 @@
+				 * TODO: iterate through resources once block transfer
+				 * is handled correctly
+				 */
+				if (pos + len < size) {
```
Done

Commit:
misc/byteorder: add support for __bswap_64

We currently support converting from cpu format to BE for
u16_t and u32_t.  Let's add u64_t as well.

NOTE: This will be used in LWM2M subsys later.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: zoap: use message id for reply matching

In the 08 Feb 2017 V1.0 LwM2M specification page 80 mentions: in
response to a "Notify" operation for which it is not interested in
any more, the LwM2M Server can send a "Reset Message".

Leshan server sends this CoAP RST response and it does not contain
the originating message token (which is also how the packet flow looks
on page 81 of the LwM2M spec).  Using the current ZoAP sources, the
client has no way of matching back to observation which needs to be
cancelled.

Let's add a match for message ID of a reply where there is no token
to handle this case.

Signed-off-by: Michael Scott <michael.scott@linaro.org>
[ricardo.salveti@linaro.org: Handle both piggybackend and separate
response (id doesn't need to match, only token).]
Signed-off-by: Ricardo Salveti <ricardo.salveti@linaro.org>

Commit:
net: lwm2m: initial library support for LWM2M

Origin: SICS-IoT / Contiki OS
URL: https://github.com/sics-iot/lwm2m-contiki/tree/lwm2m-standalone-dtls
commit: d07b0bcd77ec7e8b93787669507f3d86cfbea64a
Purpose: Introduction of LwM2M client library.
Maintained-by: Zephyr

Lightweight Machine-to-Machine (LwM2M) is a protocol stack extension
of the Constrained Application Protocol (CoAP) which uses UDP
transmission packets.

This library was based on source worked on by Joakim Eriksson,
Niclas Finne and Joel Hoglund which was adopted by Contiki and then
later revamped to work as a stand-alone library.

A VERY high level summary of the changes made:
- [ALL] sources were re-formatted to Zephyr coding standards
- [engine] The engine portion was re-written due to the heavy reliance
  on ER-CoAP APIs which are not compatible to the Zephyr CoAP APIs as
  well as other Zephyr specific needs.
- [engine] All LWM2M/IPSO object data is now abstracted into resource
  data which stores information like the data type, length, callbacks
  to help with read/write.  The engine modifies this data directly (or
  makes callbacks) instead of all of the logic for this living in each
  object's code. (This wasn't scaling well as I was implementing
  changes).
- [engine] Related to the above change, I also added a generic set of
  getter/setter functions that user applications can call to change
  the object data instead of having to add getter/setting methods in
  each object.
- [engine] The original sources shared the engine's context structure
  quite extensively causing a problem with portability. I broke up the
  context into it's individual parts: LWM2M path data, input data and
  output data and pass only the needed data into each set of APIs.
- [content format read/writer] sources were re-organized into single
  .c/h files per content formatter.
- [content format read/writer] sources were re-written where necessary
  to remove the sharing of the lwm2m engine's context and instead only
  requires the path and input or output data specific to it's
  function.
- [LwM2M objects] re-written using the new engine's abstractions

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: lwm2m: initial sample for LWM2M client

This sample utilizes the new LwM2M library by setting up default
values for LwM2M device and firmware objects and then establisting
a connection to a LwM2M server (for example Leshan Demo Server) via
the registration interface.

To use QEMU for this purpose please see:
doc/subsystems/networking/qemu_setup.rst

NOTE: This sample currently does not demonstrate DTLS/bootstrap as
neither of these is supported by the LwM2M library.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: add IPSO support w/ temperature sensor object

IPSO Smart Objects are a set of template objects based on the LwM2M
object framework which are designed to represent standard hardware
such as temperature and humidity sensors or light controls.

Let's add a place for these objects to live as well as an initial
temperature sensor object.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: lwm2m: add IPSO temperature object to LWM2M client

This commit adds IPSO temperature support to the LwM2M client sample.

NOTE: A dummy value of 25C is set during initialization and does not
change.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented:
Thank you for the initial feedback / reviews!  I've pushed a V2.
V1->V2 diff: https://hastebin.com/texubuzoru.cs

CHANGES:
- Addressed review comments (Thanks @tbursztyka and @dbkinder)
- Adjusted the patch date/times so they are all sequential.  This helps display order in github
- Use isdigit() function to check for numeric chars in atou16()
- Removed extraneous copy/paste comment in lwm2m_engine_get_rd_data()
- Fixed checkpatch errors (still have warnings for typedefs)
- Fixed sample compile breakage due to latest master branch merges in net app APIs. (Thanks @rbtchc)
- Attempted to clean up if/then/while/for nesting in string_to_path(): https://github.com/mike-scott/zephyr/blob/lwm2m-new/subsys/net/lib/lwm2m/lwm2m_engine.c#L681
- Attempted to clean up if/then/while/for nesting in do_read_op(): https://github.com/mike-scott/zephyr/blob/lwm2m-new/subsys/net/lib/lwm2m/lwm2m_engine.c#L1696

DISCUSSION POINT:
- Float32 / Float64 handling needs to have a standard representation of float data.  Currently, I'm trying use these structures: https://github.com/mike-scott/zephyr/blob/lwm2m-new/include/net/lwm2m.h#L98  Perhaps there's a better option?

mike-scott commented on subsys/net/lib/zoap/zoap.c:
```diff
@@ -615,19 +615,23 @@ struct zoap_reply *zoap_response_received(
 		int age;
 
-		if (r->tkl != tkl) {
+		if ((r->id == 0) &&
```
Apparently there's some more testing to do with this patch.  It might cause a regression where we're matching erroneously.  This patch fixes a RST return from Leshan on an observer notification where the user as stopped the observer.  The RST message has an ID which matches the newly stored reply->id, but there's no referencing token.  I should probably add this information to the patch description.

jukkar replied:
> I should probably add this information to the patch description.

Yes please.

mike-scott commented:
FYI: I'm going to be travelling this next week so I won't be as responsive to questions on this PR.  But, I will check in from time to time.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -0,0 +1,417 @@
+	struct sockaddr from_addr;
+
+	/* Save the from address */
+#if defined(CONFIG_NET_IPV6)
```
I saw similar like of calls in some other function somewhere above, perhaps some refactoring could be done here and common functionality could be separated.

mike-scott replied:
Done

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -0,0 +1,417 @@
+	int ret;
+
+	/* send request */
+	pkt = net_pkt_get_tx(firmware_net_ctx, K_FOREVER);
```
It might be better to use a timeout here as using K_FOREVER can be a bit problematic. You are checking the return value so using K_FOREVER does not make sense either.

mike-scott replied:
Done.  There is no standard time to wait that I could determine from existing code.  So, I've set it to 1 second for now.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -0,0 +1,417 @@
+		return -ENOMEM;
+	}
+
+	frag = net_pkt_get_data(firmware_net_ctx, K_FOREVER);
```
ditto

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -0,0 +1,417 @@
+
+	/* hard code URI path here -- should be pulled from package_uri */
+	ret = zoap_add_option(&request, ZOAP_OPTION_URI_PATH,
+			"large-create", strlen("large-create"));
```
sizeof("large-create") - 1 would be better here as that is evaluated during compilation

mike-scott replied:
Done.  NOTE: This file will have some immediate follow-up patches once this PR is merged and the hard-coded location for URI path is removed.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -0,0 +1,417 @@
+	ret = zoap_add_option(&request, ZOAP_OPTION_URI_PATH,
+			"large-create", strlen("large-create"));
+	ret = zoap_add_option(&request, ZOAP_OPTION_URI_PATH,
+			"1", strlen("1"));
```
ditto

mike-scott replied:
Done

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -0,0 +1,417 @@
+	}
+
+	/* reset block transfer context */
+#if defined(NET_L2_BLUETOOTH)
```
This looks suspicious, shouldn't this be CONFIG_NET_L2_BLUETOOTH?
And shouldn't we check this at runtime whether the interface we are using is BT or not? We support multiple interfaces, although I am not sure how well they work atm.

mike-scott replied:
Nice catch on the config typo.  Fixing that.

I wasn't sure how to check the interface's device type, so I used this check.  Is there a good example of this elsewhere in Zephyr?  I couldn't find any.

mike-scott replied:
Created JIRA sub-task to track this item: https://jira.zephyrproject.org/browse/ZEP-2416

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -0,0 +1,417 @@
+		k_delayed_work_init(&retransmit_work, retransmit_request);
+
+		/* start file transfer work */
+		strcpy(firmware_uri, package_uri);
```
Looks dangerous, perhaps strncpy() instead?

mike-scott replied:
Done

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_security.c:
```diff
@@ -0,0 +1,128 @@
+
+	/* initialize instance resource data */
+	INIT_OBJ_RES_DATA(res[index], i, SECURITY_SERVER_URI_ID,
+		  security_uri[index], SECURITY_URI_LEN);
```
Indentation looks wrong here and in other init macros few lines below.

mike-scott replied:
Done.  Adjusted all of the smart object INIT_OBJ_* indentations.   I had been lazy here during development as this data was changing fairly often.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_server.c:
```diff
@@ -0,0 +1,181 @@
+
+	/* initialize instance resource data */
+	INIT_OBJ_RES_DATA(res[index], i, SERVER_SHORT_SERVER_ID,
+		&server_id[index], sizeof(*server_id));
```
Indentation

mike-scott replied:
Done.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -0,0 +1,874 @@
+
+	s64_t last_update;
+
+	char ep_name[32];
```
Perhaps we could have a define instead of magic constant here?

mike-scott replied:
Done.  Added a #define for CLIENT_EP_LEN

jukkar commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -0,0 +1,874 @@
+
+static int lwm2m_rd_client_init(struct device *dev)
+{
+	k_thread_create(&lwm2m_rd_client_thread_data,
```
There are lot of threads created by lwm2m system, is it possible to combine them (or some of them)? These many threads use quite a lot of precious ram for stack space.

mike-scott replied:
I'd like to address this in a follow up patch if possible. I can add a JIRA if needed.

mike-scott replied:
Created JIRA to track memory reduction effort: https://jira.zephyrproject.org/browse/ZEP-2418

jukkar reviewed (changes requested):
Hi Mike, lot of code to review and I probably missed something. Anyway, looks good but had some comments.

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -0,0 +1,417 @@
+	ret = zoap_packet_init(&request, pkt);
+	if (ret < 0) {
+		SYS_LOG_ERR("zoap packet init error (err:%d)", ret);
+		return ret;
```
should be "goto cleanup;"

rbtchc replied:
should be "goto cleanup;"

mike-scott replied:
nice catch.  Done.

mike-scott commented:
Thank you for the reviews / feedback.  I will submit V3 on Tuesday / Wednesday (once I return from holiday):
- Better documentation
- Better patch  descriptions
- Address review comments / open JIRAs for tracking outstanding items
- Fix for leaking reply structs
- Rebase on net branch

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2482 @@
+	if (net_pkt_family(pkt) == AF_INET6) {
+		net_ipaddr_copy(&net_sin6(&from_addr)->sin6_addr,
+				&NET_IPV6_HDR(pkt)->src);
+		net_sin6(&from_addr)->sin6_port = NET_TCP_HDR(pkt)->src_port;
```
NET_TCP_HDR() -> NET_UDP_HDR()
Assuming we are using UDP.

And in latest master branch, both have been removed check 3604c391e69d253fb985aa6d072283250b454e6e

mike-scott replied:
Fixing all of these per master branch changes

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2482 @@
+	if (net_pkt_family(pkt) == AF_INET) {
+		net_ipaddr_copy(&net_sin(&from_addr)->sin_addr,
+				&NET_IPV4_HDR(pkt)->src);
+		net_sin(&from_addr)->sin_port = NET_TCP_HDR(pkt)->src_port;
```
Same here

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -0,0 +1,417 @@
+	if (net_pkt_family(pkt) == AF_INET6) {
+		net_ipaddr_copy(&net_sin6(&from_addr)->sin6_addr,
+				&NET_IPV6_HDR(pkt)->src);
+		net_sin6(&from_addr)->sin6_port = NET_TCP_HDR(pkt)->src_port;
```
same here

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -0,0 +1,417 @@
+	if (net_pkt_family(pkt) == AF_INET) {
+		net_ipaddr_copy(&net_sin(&from_addr)->sin_addr,
+				&NET_IPV4_HDR(pkt)->src);
+		net_sin(&from_addr)->sin_port = NET_TCP_HDR(pkt)->src_port;
```
same here

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
sys_cpu_to_be32() -> sys_cpu_to_be16()

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
nice catch.  done.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2482 @@
+	return NULL;
+}
+
+char *sprint_token(const u8_t *token, u8_t tkl)
```
Making this a static function as it's not used in other files

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2482 @@
+
+/* observer functions */
+
+int engine_notify_observer(u16_t obj_id, u16_t obj_inst_id, u16_t res_id)
```
Yes, I suppose this could collide (in theory).  I'll add a lwm2m_* prefix to these shared functions, even though they are only used by the subsys/net/lib/lwm2m files.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2482 @@
+	 * zoap expects that buffer->data starts at the
+	 * beginning of the CoAP header
+	 */
+	header_len = net_pkt_appdata(pkt) - pkt->frags->data;
```
Created JIRA: https://jira.zephyrproject.org/browse/ZEP-2413 to track this issue.

jukkar replied:
There is also a zoap specific Jira ticked that is related to this issue https://jira.zephyrproject.org/browse/ZEP-2210

mike-scott commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -0,0 +1,150 @@
+	bool "support for RD LWM2M client state machine"
+	default y
+	help
+	  Client will use RD client state machine to stay connected to LWM2M
```
Ack.  Providing a better description in general.

mike-scott commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -0,0 +1,150 @@
+	  This value sets the maximum number of error codes that the device
+	  object will store before ignoring new values.
+
+menu "IPSO Alliance Object Support"
```
Done.

mike-scott commented on subsys/net/lib/lwm2m/Kconfig.ipso:
```diff
@@ -0,0 +1,34 @@
+#
+
+menuconfig LWM2M_IPSO_SUPPORT
+	bool "IPSO Alliance Object Support"
```
Done.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2482 @@
+			 */
+			if (obs->event_timestamp > obs->last_timestamp &&
+			    timestamp > obs->last_timestamp +
+					(obs->min_period_sec * MSEC_PER_SEC)) {
```
Done.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2482 @@
+			 * - current timestamp > last_timestamp + max_period_sec
+			 */
+			} else if (timestamp > obs->last_timestamp +
+					(obs->max_period_sec * MSEC_PER_SEC)) {
```
Done.

mike-scott commented:
Hi all,

I've posted V3 changes for this PR.  I apologize for the sizable difference from V2.  There was a lot of renaming and refactoring due to review comments.
V2->V3 diff: https://hastebin.com/ukikeregoz.cs (LwM2M changes only, not rebase related)

HIGH LEVEL CHANGES FOR V3:
- Rebased on master branch today (7/20)
- Addressed review comments and added JIRA sub-tasks to https://jira.zephyrproject.org/browse/ZEP-216 for those items not fixed.
- Added better patch descriptions.  I think I missed a few at the beginning, I'll get those in V4.
- Fixed up some warnings when compiling the sample w/ various combinations of IPV4 / IPV6
- Added dummy temperature data to the sample client
- Re-added float content formatter for OMA TLV which only handles the whole number portions currently.  Will need to add calculations to handle the decimal portion at some point in the future (post merge).  This is useful as the temperature object can now at least show some data.
- Updated execute methods to handle return values (code provided by @rsalveti )
- Correctly free up zoap_reply and zoap_pending indexes when errors happen in net_context_sendto()
- Several master branch changes were fixed such as the removal of NET_UDP_HDR/NET_TCP_HDR macros and zoap lib changes.
- Some duplicate code was replaced with common use functions
- Removed check_perm() function as it was only used one in a place where later the permission was checked anyway.

OPEN QUESTIONS:
- Float32 / Float64 handling needs to have a standard representation of float data. Currently, I'm trying use these structures: https://github.com/mike-scott/zephyr/blob/lwm2m-new/include/net/lwm2m.h#L98 Perhaps there's a better option?

OPEN JIRA SUB-TASKS:
https://jira.zephyrproject.org/browse/ZEP-2413 : Fix LWM2M header calculation in lwm2m_engine.c
https://jira.zephyrproject.org/browse/ZEP-2416 : Fix LwM2M object firmware pull block transfer mode to select size via interface type (BT, ethernet, etc)
https://jira.zephyrproject.org/browse/ZEP-2417 : Implement DTLS/bootstrap for LwM2M library / sample
https://jira.zephyrproject.org/browse/ZEP-2418 : Reduce the overall memory usage of the LwM2M library

- Mike

mike-scott commented:
V4 update specifically addresses the ZoAP id/token matching commit "net: zoap: use message id for reply matching"
Diff: https://hastebin.com/erehohipup.php

NOTES:
- The new patch avoids matching on id 0 and breaks out the matching checks so that they happen in the right order.
- Also added a nice commit description explaining the issue that the commit is fixing.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2516 @@
+	token = zoap_header_get_token(&response, &tkl);
+	pending = zoap_pending_received(&response, zpendings, num_zpendings);
+	if (pending) {
+		/* If necessary cancel retransmissions */
```
Is this to be implemented or just missing atm? Please add "TODO"  here if this needs to be implemented later.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -0,0 +1,2516 @@
+			if (r < 0) {
+				SYS_LOG_ERR("Request handler error: %d", r);
+			} else {
+				r = net_context_sendto(pkt2, &from_addr,
```
BTW, if we want to have transparent DTLS support, then the net-app API net_app_send_pkt() should be used here, instead of direct call to net_context_sendto(). This can be changed in later PR of course.

jukkar reviewed (commented):
Looks really good. I had couple of comments but those can be done later. So if there is no objections from anyway this could be commited.

mike-scott commented:
I'm going to add a V5 which:
- Addresses your last comments
- Adds a patch to add an entry into MAINTAINERS file for subsys/net/lib/lwm2m
- Add s a patch to add an entry in doc/LICENSE.rst identifying the content formatter files and rd client files which have 3-clause BSD license.

And I'll add a JIRA subtask to track net app API conversion (send pkt)

mike-scott commented:
Apparently if the license is in the source files, then an entry in doc/LICENSING.rst is not needed (see subsys/net/ip/rpl.c)

Commit:
net: lwm2m: add SPX Apache-2.0 license tag w/ Linaro copyright

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
MAINTAINERS: add maintainer for LwM2M library / samples

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
CODEOWNERS: add entries for LwM2M library / samples

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented:
Patchset V5 diff: https://hastebin.com/ajicibemij.diff

NOTE: I'm seeing a page fault currently when testing the LwM2M client on qemu x86.  I'm currently debugging why CONFIG_X86_STACK_PROTECTION is causing this.

mike-scott commented:
FYI: There were 3 edits to fix checkpatch warnings for long lines not mentioned.

mike-scott commented:
CONFIG_X86_STACK_PROTECTION issue is related to patches recently introduced by @andrewboie.  He's aware of the issue and will post a PR to fix out of date code in include/misc/stack.h which is used when stack analyze functions are enabled.  It's not an LwM2M related issue and this PR can proceed.

mike-scott commented:
Posting V6: I found 2 calls to k_thread_create using the wrong variable for the size calculation.

mike-scott commented:
V6 diff: https://hastebin.com/xofavisepu.m

The only change for V6 was to fix the variables used in K_THREAD_STACK_SIZEOF() macro for 2x k_thread_create() calls.

andrewboie commented:
PR #907 should fix any page faults due to the net_stack_analyze() code using the wrong memory location for stack buffers

jukkar commented:
Trickle test failures should be fixed by #912

jukkar commented:
@nashif Can you force merge these changes, shippable complains about too long lines but those are from external source files that we should probably not touch.

mike-scott commented:
Actually, all of the complaints are centered around 2 patch descriptions which I can mostly fix.  1 of the complains is a source URL which is too long.. which I probably won't fix.

mike-scott commented:
Pushed V7 with only patch description changes to fix SHIPPABLE warnings (I hope).

jukkar commented:
@nashif We could merge this now but @tbursztyka earlier change request prevents merging, can you override that and merge this PR (Tomasz is on vacation right now)?

mike-scott commented:
@nashif mentioned on IRC that the "net: lwm2m: initial library support for LWM2M" commit is missing a block in the commit description similar to:

Origin: SICS-IoT / Contiki OS
URL: https://github.com/sics-iot/lwm2m-contiki/tree/lwm2m-standalone-dtls
commit: d07b0bcd77ec7e8b93787669507f3d86cfbea64a
Purpose: Introduction of LwM2M client library.
Maintained-by: Zephyr

However, this is only true for the 9 files with 3-clause BSD licenses still attached.  How should I clarify that?

mike-scott commented:
Pushed V8 to address the contribution guide requirement for the above mentioned block in the "net: lwm2m: initial library support for LWM2M" patch.  That was the only change.

mike-scott commented:
Correction: I also rebased on today's master branch.

mike-scott commented:
Ugh, SHIPPABLE complaining about the URL line length in the new change to the patch description:
https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/2782/3/tests

NOTE: running gitlint in my local repo does not show this error. :/

jukkar commented:
@nashif Could you force merge this as shippable complains about non-issue?

dbkinder reviewed (approved):
+1 for doc so far, but_lwm2m-client-sample is still empty :(
